xcmailr-webapp/src/main/java/controllers/MessageListener.java

```diff
@@ -82,6 +89,85 @@ public boolean accept(String from, String recipient)
         return false;
     }
 
+    /**
+     * Checks if forwarding the mail could trigger a loop Checks if the return-path header is empty Checks if the
+     * References/ In-Reply-To Header include a mail from this domain Checks if the custom X-Loop Header exists and
+     * checks it's content The number of hops/ to many Received fields is checked by a internal library
+     * 
+     * @param mail
+     *            the mail to check
+     * @return null if there was no loop, an error message if there was
+     * @throws MessagingException
+     */
+    private String checkForLoop(MimeMessage mail) throws MessagingException
+    {
+        String errorMessage;
+
+        // check the first return-path header
+        String[] returnPathHeaders = mail.getHeader("Return-Path");
+        String returnPathHeader = returnPathHeaders[0];
+        if (returnPathHeader.equals("") || returnPathHeader.equals("<>") || returnPathHeader.equals("< >"))
+        {
+            // loop detected;
+            errorMessage = "Return-Path is empty";
+            return errorMessage;
+        }
+       
+        // check custom X-Loop Header       
+        String customHeader = mail.getHeader(loopHeader, "###");
+        
+        if (customHeader != null)
+        {
+            customHeader = customHeader.toLowerCase();
+            String shouldBeContent = loopHeaderContent.toLowerCase();
+
+            if (customHeader.contains(shouldBeContent))
+            {
+                // loop detected;
+                errorMessage = "X-Loop header with this email adress present";
+                return errorMessage;
+            }
+        }
+
+        String id = mail.getMessageID();
+
+        String[] splitString = id.split("@");
+        String domain = splitString[1];
+        domain = domain.toLowerCase();
+
+        // check References and In-Reply-To Header
+        String referenceHeaders = mail.getHeader("References", "###");
+        if (referenceHeaders != null)
+        {
+            referenceHeaders = referenceHeaders.toLowerCase();
+            if (referenceHeaders.contains("@" + domain))
+            {
+                // loop detected;
+                errorMessage = "References field references the domain of this email adress: " + domain;
+                return errorMessage;
+            }
+        }
+
+        String inReplyToHeader = mail.getHeader("In-Reply-To", "###");
+        if (inReplyToHeader != null)
+        {
+            inReplyToHeader = inReplyToHeader.toLowerCase();
+            if (inReplyToHeader.contains("@" + domain))
+            {
+                // loop detected;
+                errorMessage = "In-Reply-To field mentions the domain of this email adress: " + domain;
```

Comment:
hence you are always just filling the errorMessage variable with some text and then immediately return it, it may be okay to return the text directly